Neos.Flow/Classes/ObjectManagement/Proxy/ProxyConstructorGenerator.php

```diff
@@ -87,14 +95,66 @@ protected function buildAssignMethodArgumentsCode(): string
+     * name of a potential non-proxied parent class of the original class.
+     *
+     * The context where the parent:: call is made is a proxy class. Therefore, the call will
+     *
+     *  - either call the method in the original class of the current proxy, if it exists, or
+     *  - call the original method in the parent class or
+     *  - call the proxied method in the parent class or
+     *  - call the original method in a subclass or
+     *  - call the proxied method in a subclass
+     *
+     * See the ProxyCompilerTest functional tests for examples of these cases.
+     *
      * @psalm-param class-string $fullClassName
      */
     protected function buildCallParentMethodCode(string $fullClassName, string $methodName): string
     {
-        # Note that the parent class "parent::" calls here is actually the original class, which is $fullClassName
-        if (!method_exists($fullClassName, $methodName)) {
+        $parentClassName = get_parent_class($fullClassName);
+        if (
+            !method_exists($fullClassName, $methodName) &&
+            ($parentClassName === false || !method_exists($parentClassName, $methodName))
+        ) {
             return '';
         }
         return "parent::{$methodName}(...\$arguments);" . PHP_EOL;
     }
+
+    /**
+     * Build code which enforces the original visibility of the constructor.
+     *
+     * This code is added to the beginning of the constructor body of a proxy class if
+     * the original visibility was not "public".
+     *
+     * For private or protected constructors there are two cases which are allowed:
+     *
+     * 1) The constructor is called from within the original class itself
+     * 2) The constructor is called from within a subclass of the original class
+     *
+     * In all other cases, an exception is thrown which is similar to the fatal error
+     * PHP would throw in that case.
+     */
+    private function buildEnforceVisibilityCode(string $originalVisibility): string
+    {
+        if ($originalVisibility === self::VISIBILITY_PUBLIC) {
+            return '';
+        }
+        $originalVisibilityString = $originalVisibility === self::VISIBILITY_PROTECTED ? 'protected' : 'private';
+        $fullOriginalClassNameWithSuffix = $this->fullOriginalClassName . Compiler::ORIGINAL_CLASSNAME_SUFFIX;
+        return <<<PHP
+        \$backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+        if (isset(\$backtrace[1]) &&
+            !is_subclass_of(\$backtrace[1]['class'], \\{$this->fullOriginalClassName}::class) &&
+            !is_subclass_of(\\{$this->fullOriginalClassName}::class, \$backtrace[1]['class']) &&
+            \$backtrace[1]['class'] !== '{$this->fullOriginalClassName}' &&
+            \$backtrace[1]['class'] !== '{$fullOriginalClassNameWithSuffix}'
+        ) {
+            throw new \\Error('Call to {$originalVisibilityString} {$this->fullOriginalClassName}::__construct() from invalid context', 1686153840);
+        }
```

Comment:
AFAIR `debug_backtrace` is not very performant (I didn't double check!) and I know, this is only done if there's a proxy class, so it might not be an issue. But I wonder: Do we really need to check whether a private constructor is invoked from the "outside"?
I mean, static analyzers and the IDE will scream at you, if you do – and if you really _do_ want to invoke the constructor, you could probably still circumvent the check with some reflection magic, anyways

Reply:
I think it's not super important to have this check (even though I worked hard to get this running). My motivation is mostly that proxy classes should not change the behavior of the proxied code if at all possible, so don't get bad surprises. 

Two made up scenarios:

1. not using or ignoring your IDE, you program something calling a private constructor which works fine because the class is proxied due to a security advice. When the advice does not match / is removed some day, this breaks, and you'll only notice it during runtime
2. a user instantiates a class in their code; the maintainer of that class at some point decides to make the constructor private, but the user won't notice that because it still works

🤷‍♂️

Reply:
yes, very good points – And doctrine uses the same kind of checks in their [deprecations](https://github.com/doctrine/deprecations/blob/eac01fbdf4e4a3635169b056608730ec0930e188/lib/Doctrine/Deprecations/Deprecation.php#L129) package, so the performance hit is probably not too bad